src/fileformat/pfsinrgbe.1
.TH "pfsinrgbe" 1
.SH NAME
pfsinrgbe \- Load images or frames in Radiance RGBE format
.SH SYNOPSIS
.B pfsinrgbe
[--linear] (<file> [--frames <range>] [--skip-missing])  [<file>...]

.SH DESCRIPTION
.I pfsinrgbe
command loads images in Radiance RGBE format and writes \fIpfs\fR
stream to the Standard Output. The \fIpfs\fR stream is usually piped
to another program for further processing. To detect the format
automatically based on the extension, use \fIpfsin\fR command.
.PP
Since RGBE is an HDR format, 'LUMINANCE' tag is set to 'RELATIVE'
(relative linear luminance values).
.PP
To read images from standard input use a single dash '-' instead of
filename. The images are read until EOF is reached.
.PP
Each file can contain a \%%d pattern, which is substituted with frame
numbers. The pattern has the same syntax as C
.I printf
command. For example, you can use \%%04d to make the frame number
four digit with proceedings zeros. You can select the frames using the
following options (the options must be always given after the file
name):
.TP
.B \--frames <range>
Range is given in mathlab / octave format:

.B "startframe:step:endframe"

Frame numbers start with
.B "startframe"
(default 0), are increased by
.B "step"
(default 1) and stop at
.B "endframe"
You can skip one of those values, for example
.I "1:100"
for frames 1,2,...,100 and
.I 0:2:
for frame 0,2,4,... up to the last file that exists.

.TP
.B \--skip-missing
Skip up to ten frames in a row if corresponding files are
missing. Otherwise the program stops reading sequence at the first
file that does not exists. This switch does not apply to the first
frame in a sequence. This switch can be useful if there is a rendered
animation where some of the frame has not been generated.

.TP
.B \--linear
Ignored for compatibility with \fIpfsinppm\fR.

.TP
.B \--radiance, -r
Correct stored values so that the luminance is reported the same in pfsview and radiance tools (xview or Photospheare). See IMPORTANT NOTE below.

.TP
.B \--quiet, -q
Disable warning message about the change in handling absolute values in RGBE files. See IMPORTANT NOTE below.

.SH EXAMPLES
.TP
pfsinrgbe intro\%%d.hdr frame\%%d.hdr | ...

Concatenate two animations into one and pass it to pipe for further
processing.

.SH IMPORTANT NOTE

There is an incompatibility between viewers of RGBE format in terms of
absolute luminance values. The ximage viewer from Radiance package
expects radiance values in RGBE files and multiplies the resulting luminance by the WHITE_EFFICACY
constant equal to 179.0f to convert from radiance to luminance. 
.PP
To remain compatible with Radiance (xview and Photospheare), 
pfstools prior to 1.9.0 divided red, green and blue values by 179 on writing and multiplied by 
that constant on reading RGBE files. This in turn caused incompatibility with other software, 
such as Photoshop or HDR toolbox. Starting from 
pfstools 1.9.0 this division / multiplication step was removed to make pfstools compatible in 
terms of absolute values with the majority of the software. The downside is that this change made 
.hdr files created with the earlier versions of pfstools incompatible (in terms of absolute 
luminance units) with the pfstools 1.9.0 or later. To avoid any mistakes because of that change, 
a long warning messages is shown each time pfsinrgbe or pfsoutrgbe is called. The message can be 
suppressed with --quiet option.
.PP
To enable the old functionality and read or write RGBE files with the correction for WHITE_EFFICACY, add --radiance option to the command line.

.SH BUGS
Please report bugs and comments on implementation to 
the discussion group http://groups.google.com/group/pfstools
.SH "SEE ALSO"
.BR pfsin (1),
.BR pfsout (1)




NEWS
This file is not maintained, but required by automake. Visit pfstools
web page to find the latest news:

http://pfstools.sourceforge.net/



src/fileformat/pfsinrgbe.cpp
/**
 * @brief Read files in Radiance's RGBE format
 * 
 * This file is a part of PFSTOOLS package.
 * ---------------------------------------------------------------------- 
 * Copyright (C) 2003,2004 Rafal Mantiuk and Grzegorz Krawczyk
 * 
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 * ---------------------------------------------------------------------- 
 * 
 * @author Grzegorz Krawczyk, <krawczyk@mpi-sb.mpg.de>
 * @author Rafal Mantiuk, <mantiuk@mpi-sb.mpg.de>
 *
 * $Id: pfsinrgbe.cpp,v 1.5 2014/06/17 21:57:08 rafm Exp $
 */

#include <config.h>

#include <cstdlib>

#include <iostream>

#include <getopt.h>
#include <pfs.h>

#include "rgbeio.h"

#define PROG_NAME "pfsinrgbe"

class QuietException 
{
};

void printHelp()
{
  std::cerr << PROG_NAME " [--linear] [--radiance] [--verbose] [--help]" << std::endl
            << "See man page for more information." << std::endl;
}


void readFrames( int argc, char* argv[] )
{
  pfs::DOMIO pfsio;

  bool verbose = false;
  bool opt_linear=false;
  bool radiance_compatibility = false;
  bool quiet = false;
  
  // Parse command line parameters
  static struct option cmdLineOptions[] = {
    { "help", no_argument, NULL, 'h' },
    { "verbose", no_argument, NULL, 'v' },
    { "linear", no_argument, NULL, 'l' },
	{ "radiance", no_argument, NULL, 'r' },
	{ "quiet", no_argument, NULL, 'q' },
	{ NULL, 0, NULL, 0 }
  };
  static const char optstring[] = "lhvrq";
    
  pfs::FrameFileIterator it( argc, argv, "rb", NULL, stdin,
    optstring, cmdLineOptions );
    
  int optionIndex = 0;
  while( 1 ) {
    int c = getopt_long (argc, argv, optstring, cmdLineOptions, &optionIndex);
    if( c == -1 ) break;
    switch( c ) {
    case 'h':
      printHelp();
      throw QuietException();
    case 'v':
      verbose = true;
      break;
	case 'r':
		radiance_compatibility = true;
		break;
	case 'q':
		quiet = true;
		break;
	case 'l':
      std::cerr << PROG_NAME << " warning: linearize option ignored for an HDR input!"
                << std::endl;
      break;
    case '?':
      throw QuietException();
    case ':':
      throw QuietException();
    }
  }

  if (!quiet && !radiance_compatibility)
	  std::cerr << PROG_NAME << " warning: starting from pfstools 1.9.0, .hdr files are read without correcting for WHITE_EFFICIENCY,"
	  " which makes the absolute values incompatible with Radiance and previos versions of pfstools but compatible with most software."
	  " Use --radiance option to retain compatibility with Radiance and earlier versions of pfstools.  Use --quiet to suppress this message."
	  " Check the manual pages for pfsinrgbe for details." << std::endl;

  while( true ) {
    pfs::FrameFile ff = it.getNextFrameFile();
    if( ff.fh == NULL ) break; // No more frames

    RGBEReader reader( ff.fh, radiance_compatibility );

    VERBOSE_STR << "reading file (linear) '" << ff.fileName << "'" << std::endl;

    pfs::Frame *frame = pfsio.createFrame( reader.getWidth(),
					 reader.getHeight() );
    pfs::Channel *X, *Y, *Z;
    frame->createXYZChannels( X, Y, Z );

    //Store RGB data temporarily in XYZ channels
    reader.readImage( X, Y, Z );
    pfs::transformColorSpace( pfs::CS_RGB, X, Y, Z, pfs::CS_XYZ, X, Y, Z );

    frame->getTags()->setString("LUMINANCE", "RELATIVE");
    const char *fileNameTag = strcmp( "-", ff.fileName )==0 ? "stdin" : ff.fileName;
    frame->getTags()->setString( "FILE_NAME", fileNameTag );

    pfsio.writeFrame( frame, stdout );
    pfsio.freeFrame( frame );

    it.closeFrameFile( ff );
  }
}


int main( int argc, char* argv[] )
{
  try {
    readFrames( argc, argv );
  }
  catch( pfs::Exception ex ) {
    fprintf( stderr, PROG_NAME " error: %s\n", ex.getMessage() );
    return EXIT_FAILURE;
  }        
  catch( QuietException ) {
    return EXIT_FAILURE;
  }        
  return EXIT_SUCCESS;
}
